cpp/unreal_plugins/SimulationController/Source/SimulationController/NavMesh.h

```diff
@@ -19,7 +20,10 @@ class NavMesh
     void cleanUpObjectReferences();
 
     std::vector<uint8_t> getRandomPoints(int num_points);
-
+    std::vector<uint8_t> getReachablePoints(const std::vector<std::vector<float>>& start_points);
```

Comment:
I think for consistency it would be better if all of these functions in our `NavMesh` class used `std::vector<float>` for input and output. When I was initially designing the interface to `getRandomPoints()`, I was trying to be consistent with our existing gym interface, which uses `std::vector<uint8_t>`. But my initial thinking was misguided. My current thinking is as follows.

- First, our gym interface has the concept of spaces for describing the shape of input and output data. So it is always possible for surrounding Python code to correctly decode an output `std::vector<uint8_t>` without needing to know any internal C++ implementation details, based on the information contained in the corresponding space. But it feels like too much work to implement a different `getWhateverSpace()` function for every scene querying function we want (e.g., `getRandomPointsSpace()`, `getReachablePointsSpace()`, etc). However, without these `getWhateverSpace()` functions, surrounding Python code needs to know too much about internal C++ implementation details to decode the data correctly. We can see evidence of this unpleasantly tight coupling in my `Env.get_random_points()` function, which needs to know somehow that the underlying C++ code returns `np.float64` data, even though this contract is not specified anywhere in the interface, and you literally need to comb through the C++ implementation to figure that out.
- Second, we need the flexibility of `std::vector<uint8_t>` in our gym interface because our agents are totally general, supporting totally heterogeneous input and output data. But we don't need this flexibility in our scene querying interfaces because the shape of the input and output data is typically known ahead of time.
- Third, I believe there is a performance penalty for sending anything other than `std::vector<uint8_t>` using `rpclib`, but I don't think that is a big deal. Our scene querying interfaces will probably not be used in tight loops, so I think it is ok to have good enough performance in exchange for cleaner interfaces. And we can always implement a fast `std::vector<uint8_t> getRandomPointsAsUint8()` function if we really need maximum performance for some reason.

Reply:
- `getRandomReachablePoints`
- `getPaths()`

Reply:
100% agree with the reasoning. For `getPaths()`, do you suggest we make both input and output `vector<vector>`? For simplicity, I think just the output being `vector<vector>>` should be sufficient.

Reply:
I agree with you. The only reason for `getPaths()` to output a `vector` of `vectors` is because the output is "ragged", i.e., each path in the output can have a different number of points, and therefore there is no way to model the output as a dense 2D array. (If we wanted to output all the points in a single dense `vector` , we would also need to output a second `vector` with the number of points in each path, but this seems like too much effort. Since the input is not ragged, I think it is sufficient for the input to be a single vector.